backend/tests/v5/e2e/routes/teamspaces/roles.test.js

```diff
@@ -0,0 +1,214 @@
+		users,
+		teamspace,
+		roles,
+	};
+};
+
+const setupData = async ({ users, teamspace, roles }) => {
+	await ServiceHelper.db.createTeamspace(teamspace, [users.tsAdmin.user]);
+
+	await Promise.all([
+		ServiceHelper.db.createUser(users.tsAdmin, [teamspace]),
+		ServiceHelper.db.createUser(users.normalUser, [teamspace]),
+		ServiceHelper.db.createUser(users.nobody, []),
+		ServiceHelper.db.createRoles(teamspace, roles),
+	]);
+};
+
+const testGetRoleList = () => {
+	const basicData = generateBasicData();
+	const { users, teamspace, roles } = basicData;
+	beforeAll(async () => {
+		await setupData(basicData);
+	});
+
+	const route = (key, ts = teamspace) => `/v5/teamspaces/${ts}/roles${key ? `?key=${key}` : ''}`;
+
+	describe('Get role list', () => {
+		describe.each([
+			['user does not have a valid session', undefined, undefined, false, templates.notLoggedIn],
+			['teamspace does not exist', users.tsAdmin.apiKey, generateRandomString(), false, templates.teamspaceNotFound],
+			['user is not a member of the teamspace', users.nobody.apiKey, undefined, false, templates.teamspaceNotFound],
+			['user is a ts admin', users.tsAdmin.apiKey, undefined, true],
+			['user is a member', users.normalUser.apiKey, undefined, true],
+		])('', (desc, key, ts, success, expectedRes) => {
+			test(`should ${success ? 'succeed' : `fail with ${expectedRes.code}`} if ${desc}`, async () => {
+				const expectedStatus = success ? templates.ok.status : expectedRes.status;
+				const res = await agent.get(route(key, ts)).expect(expectedStatus);
+				if (success) {
+					expect(res.body.roles).toEqual(expect.arrayContaining(roles));
+				} else {
+					expect(res.body.code).toEqual(expectedRes.code);
+				}
+			});
+		});
+	});
+};
+
+const testAddRole = () => {
+	const basicData = generateBasicData();
+	const { users, teamspace } = basicData;
+
+	beforeAll(async () => {
+		await setupData(basicData);
+	});
+
+	const roleData = {
+		name: ServiceHelper.generateRandomString(),
+		color: DEFAULT_ROLES[0].color,
+		users: [users.tsAdmin.user],
+	};
+
+	const getRolesRoute = (key, ts = teamspace) => `/v5/teamspaces/${ts}/roles${key ? `?key=${key}` : ''}`;
+	const addRoleRoute = (key, ts = teamspace) => `/v5/teamspaces/${ts}/roles${key ? `?key=${key}` : ''}`;
+
+	describe.each([
+		['user does not have a valid session', undefined, undefined, roleData, false, templates.notLoggedIn],
+		['teamspace does not exist', users.tsAdmin.apiKey, generateRandomString(), roleData, false, templates.teamspaceNotFound],
+		['user is not a member of the teamspace', users.nobody.apiKey, undefined, roleData, false, templates.teamspaceNotFound],
+		['user is not a teamspace admin', users.normalUser.apiKey, undefined, roleData, false, templates.notAuthorized],
+		['user is a ts admin and data is valid', users.tsAdmin.apiKey, undefined, roleData, true],
+		['data is invalid (missing name)', users.tsAdmin.apiKey, undefined, { ...roleData, name: undefined }, false, templates.invalidArguments],
```

Comment:
Can we add name clash as a test case

Reply:
Can we also add a test that post 2 roles with the same name pretty much in parallel and make sure one of them fails (to test the unqiue index thing)